### PR TITLE
[PATCH v3] linux-gen: pktio: dpdk: add dpdk packet parser

### PR DESCRIPTION
**Pull request by bogdanPricope**
Add dpdk packet parser.

### CODE REVIEW

File: platform/linux-generic/pktio/dpdk.c [hunk 1]
```diff
@@ -549,7 +553,7 @@ static inline int mbuf_to_pkt(pktio_entry_t *pktio_entry,
 		if (pktio_cls_enabled(pktio_entry))
 			copy_packet_cls_metadata(&parsed_hdr, pkt_hdr);
 		else if (parse_layer != ODP_PROTO_LAYER_NONE)
-			packet_parse_layer(pkt_hdr, parse_layer);
+			dpdk_packet_parse_layer(pkt_hdr, mbuf, parse_layer);
```
Comment (lumag) [inline platform/linux-generic/pktio/dpdk.c]: I have been thinking about similar change in generic code. However in generic code some PktIOs call parsing function using buffer, which is not a part of the packet (yet). So it is not possible to always pass odp_packet_t/odp_buffer_t to parsing functions.

**Comment by bogdanPricope** [comment]
@lumag  Sorry, I don't get your point. Can you elaborate your comment (what len, what parse flags) ?

**Comment by bogdanPricope** [comment]
@lumag  Yes,I liked it too... yet, was dropped due review (not changing all pktios makes sense, also).

**Comment by lumag** [comment]
@bogdanPricope what kind of response did it receive? I don't see any comment here.

**Comment by bogdanPricope** [comment]
@lumag  It was in an 'ODP Arch Sync' meeting: suggestion from Bill was to avoid adding repetitive code sequence. (packet_parse_reset(&parsed_hdr);packet_set_len(&parsed_hdr, pkt_len);)  but add this flag to parse or not internally.

**Comment by lumag** [comment]
Reviewed-by: Dmtiry Eremin-Solenikov

**Comment by matiaselo** [comment]
I run quick tests with odp_l2fwd (dpdk-zero-copy + disable-abi-compat, odp_l2fwd -a 0 -i 0,1 -m 0 -e 1 -k 1 -c N) and with these patches the performance is on average ~4% lower than before. I'm using Intel XL710 NICs.

File: platform/linux-generic/include/odp_packet_dpdk.h [hunk 2]
```diff
@@ -72,4 +72,15 @@ typedef struct ODP_ALIGNED_CACHE {
 	dpdk_opt_t opt;
 } pkt_dpdk_t;
 
+/** Packet parser using DPDK interface */
+int dpdk_packet_parse_layer(odp_packet_hdr_t *pkt_hdr,
```
Comment (matiaselo) [inline platform/linux-generic/include/odp_packet_dpdk.h]: Should be inlined.

**Comment by bogdanPricope** [comment]
@matiaselo  I am not really surprised about this 4% drop as it implements more features than current parser: drop packets on ipv4/ipv6/udp/tcp errors is not implemented in current parser (we are not comparing the same things). More, the gain on using rte_mbuf packet_type and ol_flags is not so big, while moving csum tests inside parser is not removing many conditionals. 

Point is, this parser is making code cleaner, cleanly extensible with dpdk calls (if needed) and add the missing feature; it may be useful for odp-dpdk as well.

And yes, the drop can be recuperated with inlines and other tricks.

**Comment by muvarov** [comment]
lets discuss this PR on Monday one more time.

**Comment by matiaselo** [comment]
>Point is, this parser is making code cleaner, cleanly extensible with dpdk calls (if needed) and add the missing feature

This is all good and I'm not against merging this patch set. I'm just pointing out there are still optimizations to be made. I can live with the small performance degradation for now, as it doesn't effect raw throughput when packet parser is disabled.

The commit log for `linux-gen: pktio: dpdk: add specific packet parser` patch is misleading as it only states that parsing performance is increased.

> it may be useful for odp-dpdk as well.

This should definitely be implemented for odp-dpdk as well.

**Comment by bogdanPricope** [comment]
Technically, only third patch (linux-gen: pktio: dpdk: integrate csum with packet parser) is reducing performance… and for that patch I already updated the commit log.

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**